// linq-to-typescript/commonjs/sync/static/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.repeat = exports.range = exports.from = exports.flatten = exports.enumerateObject = exports.empty = void 0;
var empty_1 = require("./empty");
Object.defineProperty(exports, "empty", { enumerable: true, get: function () { return empty_1.empty; } });
var enumerateObject_1 = require("./enumerateObject");
Object.defineProperty(exports, "enumerateObject", { enumerable: true, get: function () { return enumerateObject_1.enumerateObject; } });
var flatten_1 = require("./flatten");
Object.defineProperty(exports, "flatten", { enumerable: true, get: function () { return flatten_1.flatten; } });
var from_1 = require("./from");
Object.defineProperty(exports, "from", { enumerable: true, get: function () { return from_1.from; } });
var range_1 = require("./range");
Object.defineProperty(exports, "range", { enumerable: true, get: function () { return range_1.range; } });
var repeat_1 = require("./repeat");
Object.defineProperty(exports, "repeat", { enumerable: true, get: function () { return repeat_1.repeat; } });
